### PR TITLE
Fix race condition between gc and committing

### PR DESCRIPTION
**Pull request by csmarchbanks**
Fixes: https://github.com/prometheus/tsdb/issues/230
Perhaps @gouthamve can review?

Signed-off-by: Chris Marchbanks <csmarchbanks@gmail.com>

### CODE REVIEW

**Comment by krasi-georgiev** [comment]
so we are deleting the series and than on Commit adding these again. 
I am still looking through the code, but can we try to just make the `gc` smarter and maybe keep reference to these until the next commit.
maybe use 
https://github.com/prometheus/tsdb/pull/378/files#diff-0ced3454d447442a96a77b33ebe87300R1198

**Comment by csmarchbanks** [comment]
I considered that, but it seemed a little strange to be doing deletes on a Commit, and the gc path would be harder to reason about afterward.

Perhaps another option is to add a `hasPendingSamples` flag to each series when an Add is done, and only delete the series if there are no pending samples? Pending samples would then be cleared on Commit/Rollback.

Let me know what you think about those options/any other options.

**Comment by krasi-georgiev** [comment]
> Perhaps another option is to add a hasPendingSamples flag to each series when an Add is done, and only delete the series if there are no pending samples? Pending samples would then be cleared on Commit/Rollback.

that seems like a good solution. Fabian did something similar with the blocks during compaction.

**Comment by csmarchbanks** [comment]
Updated, I ended up rearranging some of the error sequencing in `AddFast` in order to only lock the series once. If that is considered backwards breaking (possible that a different error would now be returned), let me know and I will change it.

**Comment by krasi-georgiev** [comment]
it should  be fine, 
checking now.

File: head.go [hunk 1]
```diff
@@ -620,21 +620,23 @@ func (a *headAppender) Add(lset labels.Labels, t int64, v float64) (uint64, erro
...
+	s := a.head.series.getByID(ref)
 	if s == nil {
 		return errors.Wrap(ErrNotFound, "unknown series")
 	}
 	s.Lock()
 	err := s.appendable(t, v)
-	s.Unlock()
-
 	if err != nil {
+		s.Unlock()
```
Comment (krasi-georgiev) [inline head.go]: I bit more simple as:
```
if err := s.appendable(t, v); err != nil {
	s.Unlock()
	return err
}
```

File: head.go [hunk 2]
```diff
@@ -1256,9 +1264,10 @@ type memSeries struct {
 	chunkRange   int64
 	firstChunkID int
 
-	nextAt    int64 // timestamp at which to cut the next chunk.
-	lastValue float64
-	sampleBuf [4]sample
+	nextAt     int64 // timestamp at which to cut the next chunk.
+	lastValue  float64
+	sampleBuf  [4]sample
+	addPending bool
```
Comment (krasi-georgiev) [inline head.go]: a little comment of its purpose would be good  and what about if we call it `pendingCommit`

**Comment by csmarchbanks** [comment]
Thanks for the feedback! I also like this version much better. I fixed your comments and added another test.

Right now it looks like `Delete` does not have to change, but https://github.com/prometheus/tsdb/pull/270 will need to potentially clear out any pending samples when it is resurrected.

File: head.go [hunk 3]
```diff
@@ -713,6 +715,11 @@ func (a *headAppender) Commit() error {
 
 func (a *headAppender) Rollback() error {
 	a.head.metrics.activeAppenders.Dec()
+	for _, s := range a.samples {
+		s.series.Lock()
+		s.series.pendingCommit = false
+		s.series.Unlock()
+	}
```
[the remark this answers is not in the document]
Comment (gouthamve) [reply head.go]: `a.samples` will have different series and here we are acquiring each series' lock. The `s` in `s.series.Lock()` comes from looping.

**Comment by gouthamve** [comment]
LGTM after @krasi-georgiev's comments are addressed :) Thanks for taking a look at this!

File: head.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply head.go]: ooo yeah that is right.

**Comment by csmarchbanks** [comment]
Addressed all of @krasi-georgiev comments, thanks again for the review!

**Comment by krasi-georgiev** [comment]
Thanks! Job well done!

**Pull request merged**